consensus/istanbul/core/preprepare_test.go

```diff
@@ -55,85 +56,139 @@ func TestHandlePreprepare(t *testing.T) {
-						// hack: force set subject that future message can be simulated
-						c.current = newTestRoundState(
-							&istanbul.View{
-								Round:    big.NewInt(0),
-								Sequence: big.NewInt(0),
-							},
-							c.valSet,
-						)
-
-					} else {
-						c.current.SetSequence(big.NewInt(10))
+						// replica 0 is the proposer
+						c.state = StatePreprepared
 					}
 				}
 				return sys
 			}(),
+			func(_ *testSystem) istanbul.RoundChangeCertificate {
+				return istanbul.RoundChangeCertificate{}
+			},
 			makeBlock(1),
-			errFutureMessage,
+			errNotFromProposer,
 			false,
 		},
 		{
-			// non-proposer
+			// errOldMessage
 			func() *testSystem {
 				sys := NewTestSystemWithBackend(N, F)
 
-				// force remove replica 0, let replica 1 be the proposer
-				sys.backends = sys.backends[1:]
-
 				for i, backend := range sys.backends {
 					c := backend.engine.(*core)
 					c.valSet = backend.peers
 					if i != 0 {
-						// replica 0 is the proposer
 						c.state = StatePreprepared
+						c.current.SetSequence(big.NewInt(10))
+						c.current.SetRound(big.NewInt(10))
 					}
 				}
 				return sys
 			}(),
+			func(_ *testSystem) istanbul.RoundChangeCertificate {
+				return istanbul.RoundChangeCertificate{}
+			},
 			makeBlock(1),
-			errNotFromProposer,
+			errOldMessage,
 			false,
 		},
 		{
-			// errOldMessage
+			// ROUND CHANGE certificate missing
 			func() *testSystem {
 				sys := NewTestSystemWithBackend(N, F)
 
-				for i, backend := range sys.backends {
+				for _, backend := range sys.backends {
 					c := backend.engine.(*core)
 					c.valSet = backend.peers
-					if i != 0 {
-						c.state = StatePreprepared
-						c.current.SetSequence(big.NewInt(10))
-						c.current.SetRound(big.NewInt(10))
-					}
+					c.state = StatePreprepared
+					c.current.SetRound(big.NewInt(1))
 				}
 				return sys
 			}(),
+			func(_ *testSystem) istanbul.RoundChangeCertificate {
+				return istanbul.RoundChangeCertificate{}
+			},
 			makeBlock(1),
-			errOldMessage,
+			errMissingRoundChangeCertificate,
+			false,
+		},
+		{
+			// ROUND CHANGE certificate invalid, duplicate messages.
+			func() *testSystem {
+				sys := NewTestSystemWithBackend(N, F)
+
+				for _, backend := range sys.backends {
+					c := backend.engine.(*core)
+					c.valSet = backend.peers
+					c.state = StatePreprepared
+					c.current.SetRound(big.NewInt(1))
+				}
+				return sys
+			}(),
+			func(sys *testSystem) istanbul.RoundChangeCertificate {
+				// Duplicate messages
+				roundChangeCertificate := sys.getRoundChangeCertificate(t, *(sys.backends[0].engine.(*core).currentView()), istanbul.EmptyPreparedCertificate())
+				roundChangeCertificate.RoundChangeMessages[1] = roundChangeCertificate.RoundChangeMessages[0]
+				return roundChangeCertificate
+			},
+			makeBlock(1),
+			errInvalidRoundChangeCertificateDuplicate,
+			false,
+		},
+		{
+			// ROUND CHANGE certificate contains PREPARED certificate for a different block.
```

Comment:
Can we add a test case with a valid ROUND CHANGE certificate? And one with a valid ROUND CHANGE certificate that contains a valid PREPARED certificate?